<!DOCTYPE html>
<html lang="en">
<head>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta name="generator" content="Pelican" />
        <title>"Fast searching in a sorted genetics file & the curse of gzip compression"</title>
        <link rel="stylesheet" href="https://blog.llandsmeer.com/theme/css/main.css" />
        <link href="/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="The Weekend Writeup Atom Feed" />
        <meta name="description" content="As a bioinformatician, how often have you found yourself staring at the screen waiting for grep to find a certain gene or chromosome-basepair..." />
</head>

<body id="index" class="home">
        <header id="banner" class="body">
                <h1><a href="https://blog.llandsmeer.com/">The Weekend Writeup <strong>Lennart Landsmeer's Personal Blog</strong></a></h1>
                <nav><ul>
                </ul></nav>
        </header><!-- /#banner -->
<section id="content" class="body">
  <article>
    <header>
      <h1 class="entry-title">
        <a href="https://blog.llandsmeer.com/tech/2019/12/28/search-sorted.html" rel="bookmark"
           title="Permalink to "Fast searching in a sorted genetics file & the curse of gzip compression"">"Fast searching in a sorted genetics file & the curse of gzip compression"</a></h1>
    </header>

    <div class="entry-content">
<footer class="post-info">
        <abbr class="published" title="2019-12-28T21:07:00+01:00">
        28 dec 2019
        </abbr>


</footer><!-- /.post-info -->      <p>As a bioinformatician,
how often have you found yourself staring at the screen waiting for <code>grep</code>
to find a certain gene or chromosome-basepair location in a gargantuan vcf, bed or gwas file?
And then realize you forgot adding a field delimiter to your pattern so your search
included way too much results?
Thats a problem I have wanted so solve for a long time.</p>
<p>Luckily, these files are usually sorted, which means we should be able to use a binary search!
So thats what I wrote this weekend;
a tool you can use to instantly find genomic regions
in anything from vcfs, bed files, csv data, gwas to plink genetic statistics with
a friendly user interface.
As implementing a binary sort on a text file
presented me with a lot of interesting challenges, skip below for some implementation details.</p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>grep<span class="w"> </span>-E<span class="w"> </span><span class="s1">&#39;11 119[0-9]{3} &#39;</span><span class="w"> </span>data
---<span class="w"> </span><span class="m">11</span>:119854:T:G<span class="w"> </span><span class="m">11</span><span class="w"> </span><span class="m">119854</span><span class="w"> </span>T<span class="w"> </span>G<span class="w"> </span>G<span class="w"> </span>T<span class="w"> </span><span class="m">623</span>.98<span class="w"> </span><span class="m">0</span>.02301<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">624</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">1</span>.8438e-05<span class="w"> </span>-0<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span>.011331<span class="w"> </span><span class="m">1</span>.84374e-05
---<span class="w"> </span><span class="m">11</span>:119945:G:C<span class="w"> </span><span class="m">11</span><span class="w"> </span><span class="m">119945</span><span class="w"> </span>G<span class="w"> </span>C<span class="w"> </span>C<span class="w"> </span>G<span class="w"> </span><span class="m">622</span>.6<span class="w"> </span><span class="m">1</span>.3961<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">623</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span>.0011187<span class="w"> </span>-0<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span>.71666<span class="w"> </span><span class="m">0</span>.00111868
<span class="m">10</span>,98s<span class="w"> </span><span class="o">(</span><span class="m">9</span>,84s<span class="o">(</span>u<span class="o">)</span><span class="w"> </span>+<span class="w"> </span><span class="m">1</span>,14s<span class="o">(</span>s<span class="o">)</span><span class="w"> </span>4kb<span class="w"> </span><span class="m">33</span>+1<span class="w"> </span>ctx<span class="o">)</span>

$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>python3<span class="w"> </span>gensearch.py<span class="w"> </span>-f<span class="w"> </span>data<span class="w"> </span>--chr<span class="w"> </span>Chr<span class="w"> </span>--bp<span class="w"> </span>BP<span class="w"> </span>chr11:119500/1000
SNPID<span class="w"> </span>RSID<span class="w"> </span>Chr<span class="w"> </span>BP<span class="w"> </span>A_allele<span class="w"> </span>B_allele<span class="w"> </span>MinorAllele<span class="w"> </span>MajorAllele<span class="w"> </span>AA<span class="w"> </span>AB<span class="w"> </span>BB<span class="w"> </span>AA_calls<span class="w"> </span>AB_calls<span class="w"> </span>BB_calls<span class="w"> </span>MAF<span class="w"> </span>HWE<span class="w"> </span>missing<span class="w"> </span>missing_calls<span class="w"> </span>Info<span class="w"> </span>CAF
---<span class="w"> </span><span class="m">11</span>:119854:T:G<span class="w"> </span><span class="m">11</span><span class="w"> </span><span class="m">119854</span><span class="w"> </span>T<span class="w"> </span>G<span class="w"> </span>G<span class="w"> </span>T<span class="w"> </span><span class="m">623</span>.98<span class="w"> </span><span class="m">0</span>.02301<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">624</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">1</span>.8438e-05<span class="w"> </span>-0<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span>.011331<span class="w"> </span><span class="m">1</span>.84374e-05
---<span class="w"> </span><span class="m">11</span>:119945:G:C<span class="w"> </span><span class="m">11</span><span class="w"> </span><span class="m">119945</span><span class="w"> </span>G<span class="w"> </span>C<span class="w"> </span>C<span class="w"> </span>G<span class="w"> </span><span class="m">622</span>.6<span class="w"> </span><span class="m">1</span>.3961<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">623</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span>.0011187<span class="w"> </span>-0<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">0</span>.71666<span class="w"> </span><span class="m">0</span>.00111868
<span class="m">0</span>,11s<span class="w"> </span><span class="o">(</span><span class="m">0</span>,08s<span class="o">(</span>u<span class="o">)</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span>,03s<span class="o">(</span>s<span class="o">)</span><span class="w"> </span>14kb<span class="w"> </span><span class="m">0</span>+1<span class="w"> </span>ctx<span class="o">)</span>
</code></pre></div>

<!--


<div class="highlight"><pre><span></span><code>$<span class="w"> </span>python3<span class="w"> </span>gensearch.py<span class="w"> </span>-f<span class="w"> </span>stats
Missing<span class="w"> </span>--chr<span class="w"> </span>or<span class="w"> </span>--bp.<span class="w"> </span>Possible<span class="w"> </span>column<span class="w"> </span>names<span class="w"> </span>are:
SNPID<span class="w">         </span><span class="o">(</span>rs22806..,<span class="w"> </span>rs36046..,<span class="w"> </span>rs14091..,<span class="w"> </span>rs68941..,<span class="w"> </span>..<span class="o">)</span>
Chr<span class="w">           </span><span class="o">(</span><span class="m">15</span>,<span class="w"> </span><span class="m">13</span>,<span class="w"> </span><span class="m">20</span>,<span class="w"> </span>0X,<span class="w"> </span>..<span class="o">)</span>
BP<span class="w">            </span><span class="o">(</span><span class="m">5088456</span>,<span class="w"> </span><span class="m">45730252</span>,<span class="w"> </span><span class="m">32947066</span>,<span class="w"> </span><span class="m">89827248</span>,<span class="w"> </span>..<span class="o">)</span>
A_allele<span class="w">      </span><span class="o">(</span>CAG,<span class="w"> </span>AAG,<span class="w"> </span>TTA,<span class="w"> </span>GTTGT,<span class="w"> </span>..<span class="o">)</span>
B_allele<span class="w">      </span><span class="o">(</span>CATA,<span class="w"> </span>CAG,<span class="w"> </span>TTA,<span class="w"> </span>GTCCC,<span class="w"> </span>..<span class="o">)</span>
MinorAllele<span class="w">   </span><span class="o">(</span>CATA,<span class="w"> </span>CAG,<span class="w"> </span>AAG,<span class="w"> </span>TTA,<span class="w"> </span>..<span class="o">)</span>
MajorAllele<span class="w">   </span><span class="o">(</span>CAG,<span class="w"> </span>TTA,<span class="w"> </span>GTTGT,<span class="w"> </span>TC,<span class="w"> </span>..<span class="o">)</span>
<span class="o">[</span>...<span class="o">]</span>

$<span class="w"> </span>python3<span class="w"> </span>gensearch.py<span class="w"> </span>-f<span class="w"> </span>stats<span class="w"> </span>--chr<span class="w"> </span>Chr<span class="w"> </span>--bp<span class="w"> </span>BP
No<span class="w"> </span>query<span class="w"> </span>specified
Available<span class="w"> </span>regions:
<span class="m">01</span>:10177-249240543
<span class="m">02</span>:10179-243188367
<span class="m">03</span>:60069-197962381
<span class="m">04</span>:10005-191043881
<span class="m">05</span>:10043-180904689
<span class="o">[</span>...<span class="o">]</span>
</code></pre></div>


-->

<h1>Implementation</h1>
<p>Here are some interesting challenges and their solutions
for searching in sorted files. These are all genetic data agnostic,
so maybe they'll be of use in other projects as well.</p>
<h4>Reading between the lines</h4>
<p>As we perform our binary search, we'll jump from one
file position to another, until we find the line matching the user's query.
These positions are most likely not at the
start of a line. So we need some kind op method to read
one or more lines given a position in the file.
Why more than one line? Continue reading :)</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_lines_at</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">nlines</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">buffer_size</span><span class="o">=</span><span class="mi">40960</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;reads &lt;nlines&gt; complete lines from binary stream f at pos.</span>

<span class="sd">    read_lines_at() searches for the first occurance of</span>
<span class="sd">    a newline, then proceeds from there by returning the</span>
<span class="sd">    next &lt;nlines&gt; complete lines. When (the nonnegative) &lt;pos&gt; is zero,</span>
<span class="sd">    the initial search is skipped.</span>
<span class="sd">    All lines must be contained in a read of &lt;buffer_size&gt; from pos.</span>
<span class="sd">    If this does not fit, the function fails.</span>
<span class="sd">    If EOF is found as the last line, b&#39;&#39; is returned instead (like</span>
<span class="sd">    f.read() at EOF). Another read after that would result in failure.</span>

<span class="sd">    read_lines_at() returns a tuple (position of line start, [lines]).</span>
<span class="sd">    On failure, (_, None) is returned.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">buffer_size</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">data</span><span class="p">:</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="sa">b</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span> <span class="k">if</span> <span class="n">pos</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">s</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">pos</span> <span class="o">!=</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">==</span> <span class="n">buffer_size</span><span class="p">:</span>
            <span class="c1"># buffer size is not large enough</span>
            <span class="k">return</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="kc">None</span>
        <span class="n">line_start</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">s</span>
        <span class="n">e</span> <span class="o">=</span> <span class="n">s</span>
        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nlines</span><span class="p">):</span>
            <span class="n">e</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="sa">b</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">e</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">idx</span> <span class="o">&lt;</span> <span class="n">nlines</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">line_start</span><span class="p">,</span> <span class="kc">None</span>
                <span class="n">e</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">s</span><span class="p">:</span><span class="n">e</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">e</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">s</span> <span class="o">=</span> <span class="n">e</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">line_start</span><span class="p">,</span> <span class="n">lines</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pos</span><span class="p">,</span> <span class="kc">None</span>
</code></pre></div>

<h4>Binary search in a text file</h4>
<p>Next up, our binary search itself!
Remember, normally we start in the middle of a sorted array.
There, we compare our query item with that element, and
then either perform the algorithm on the region below or above the middle,
or our element is the right element and we return its index.
However, region queries usually turn up much more than one index.
So we need to find the <em>leftmost</em> match for our query, and then just read
lines until the query does not match anymore.</p>
<p>To make a binary search find this leftmost line, we need to look at two lines
at the same time. Then the 'element' our algorithm is looking for, are the two
consecutive lines where the first line does not match the query, and the second line
does. This combined with some logic for the first and last lines, I ended up with:</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">search_left</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">seek_from</span><span class="p">,</span> <span class="n">seek_to</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">keyf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;finds the first line for which keyf(at, line) == value.&#39;&#39;&#39;</span>
    <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="o">=</span> <span class="n">seek_from</span><span class="p">,</span> <span class="n">seek_to</span>
    <span class="k">while</span> <span class="n">a</span> <span class="o">!=</span> <span class="n">b</span><span class="p">:</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>
        <span class="n">line_start</span><span class="p">,</span> <span class="p">(</span><span class="n">left</span><span class="p">,</span> <span class="n">right</span><span class="p">)</span> <span class="o">=</span> <span class="n">read_lines_at</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">nlines</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">left_value</span> <span class="o">=</span> <span class="n">keyf</span><span class="p">(</span><span class="n">line_start</span><span class="p">,</span> <span class="n">left</span><span class="p">)</span> <span class="k">if</span> <span class="n">left</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">right_value</span> <span class="o">=</span> <span class="n">keyf</span><span class="p">(</span><span class="n">line_start</span><span class="p">,</span> <span class="n">right</span><span class="p">)</span> <span class="k">if</span> <span class="n">right</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">left_value</span> <span class="o">!=</span> <span class="n">value</span> <span class="ow">and</span> <span class="n">right_value</span> <span class="o">==</span> <span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">line_start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">left</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">right</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="n">right</span> <span class="ow">or</span> <span class="n">right_value</span> <span class="o">&gt;=</span> <span class="n">value</span><span class="p">:</span>
            <span class="n">b</span> <span class="o">=</span> <span class="n">pos</span>
        <span class="k">elif</span> <span class="n">right_value</span> <span class="o">&lt;</span> <span class="n">value</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">==</span> <span class="n">b</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">pos</span>
    <span class="k">if</span> <span class="n">pos</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">line_start</span><span class="p">,</span> <span class="p">(</span><span class="n">first_line</span><span class="p">,)</span> <span class="o">=</span> <span class="n">read_lines_at</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">keyf</span><span class="p">(</span><span class="n">line_start</span><span class="p">,</span> <span class="n">first_line</span><span class="p">)</span> <span class="o">==</span> <span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="n">first_line</span>
        <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span>
    <span class="k">return</span> <span class="n">seek_to</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="kc">None</span>
</code></pre></div>

<h4>Unrolled breadth first search</h4>
<p>While the lines within for a certain chromosome are
usually sorted with respect to basepair position,
the chromosomes themselves are not always sorted in a file.
So you have to poke the file in a smart way until you find
one occurancy of your query chromosome.
Let's call this the pivot line.
Then you can define a complete order on the file by
comparing a chromosome's name with the query chromsome.
If they are not equal, compare the line's position with the position of the pivot line.</p>
<p>So what is a simple and reasonable smart way of poking our file?
A breadth first search! Thinking of how to implement that on
an array was a fun excercise, but ended up being quite simple.
Performing it on a 'array' with indices continuous in the range 0-1,
the traversal would follow: 1/2; 1/4, 3/4; 1/8, 3/8, 5/8, 7/8 etc..
Then the pattern is fairly obvious.</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">bfs_indices</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">depth</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;generates linear breadth first search indices.</span>

<span class="sd">    Returns array indices as if a breadth first search</span>
<span class="sd">    is performed on a binary tree constructed from a</span>
<span class="sd">    sorted linear array of length &lt;length&gt;. Defaults</span>
<span class="sd">    to a search through the entire array, but early</span>
<span class="sd">    stopping is possible by providing the maximum</span>
<span class="sd">    tree search level &lt;depth&gt;.</span>

<span class="sd">    &gt;&gt;&gt; list(linear_bfs_indices(length=10))</span>
<span class="sd">    [5, 2, 7, 1, 3, 6, 8, 0, 1, 3, 4, 5, 6, 8, 9]</span>
<span class="sd">    &gt;&gt;&gt; list(linear_bfs_indices(length=512, depth=3))</span>
<span class="sd">    [256, 128, 384, 64, 192, 320, 448]</span>
<span class="sd">    &gt;&gt;&gt; len(set(linear_bfs_indices(length=10)))</span>
<span class="sd">    10</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">denom</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="k">if</span> <span class="n">depth</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">depth</span> <span class="o">=</span> <span class="n">length</span><span class="o">.</span><span class="n">bit_length</span><span class="p">()</span>
    <span class="k">while</span> <span class="n">depth</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">num</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">denom</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">length</span> <span class="o">*</span> <span class="n">num</span> <span class="o">//</span> <span class="n">denom</span>
        <span class="n">depth</span><span class="p">,</span> <span class="n">denom</span> <span class="o">=</span> <span class="n">depth</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">denom</span><span class="o">*</span><span class="mi">2</span>
</code></pre></div>

<h4>The curse of gzip compression</h4>
<p>Still then, there is the annoying problem that
all these neatly sorted files are usually compressed with <code>gzip</code> -
which does not allow random access halfway the file in the general case.
Thus searching can only be done by reading the file line by line,
even though there are compression tools available like <code>bgzip</code>
(or maybe <code>dictzip</code> or <code>bzip2</code>, see <a href="https://stackoverflow.com/questions/429987/compression-formats-with-good-support-for-random-access-within-archives">stackoverflow</a>)
that create a compressed file for which binary search would work perfectly,
readable on all systems by default, with only slightly larger file sizes...
<code>bgzip</code> is usually used for vcf and bed files together with <code>tabix</code>,
but by itself it already provides enough information for fast searching.</p>
<p>I did some expriments that showed that with a lot of hard work
random access to a <code>gzip</code> seems to be possible for
files compressed with the default flags.
Maybe there will be a post on that later.</p>
<p><strong>Later</strong>: so I modified <a href="https://github.com/jibsen/tinf">a public deflate implementation</a>
s.t. it would be able to start halfway a gzip file and continue reading from there,
keeping track of known and unknown output characters. More general, it doesn't only
read gzip files, but any random container format containing DEFLATE compressed data.
Under the constraint of only outputting printable ascii characters, we can recover
data after starting randomly inside a DEFLATE stream. Sadly, this only works for files
with very low complexity (like <code>seq 1 100000</code>), not for genetics files because <code>gzip</code> likes
to create arbitrary long (~100 MB's) backreferences in its output for neglible compression
gains. Anyway, if you need a program that's able to recover DEFLATE streams from unknown
container files, it's <a href="https://github.com/llandsmeer/gzip-random-seek">available here</a>.</p>
    </div><!-- /.entry-content -->

  </article>
</section>
        <section id="extras" class="body">
        </section><!-- /#extras -->

</body>
</html>